In [1]:
# importing necessary libraries
import pandas as pd

In [2]:
# Load the filtered dataset
df = pd.read_csv("../data/kickstarter_filtered.csv")

In [3]:
# Get the first 5 rows of the dataset
df.head()

,project id,name,url,category,subcategory,location,status,goal,pledged,funded percentage,backers,funded date,levels,reward levels,updates,comments,duration
0,39409,WHILE THE TREES SLEEP,http://www.kickstarter.com/projects/emiliesaba...,Film & Video,Short Film,"Columbia, MO",successful,10500.0,11545.0,1.099524,66,"Fri, 19 Aug 2011 19:28:17 -0000",7,"$25,$50,$100,$250,$500,$1,000,$2,500",10,2,30.00
1,126581,Educational Online Trading Card Game,http://www.kickstarter.com/projects/972789543/...,Games,Board & Card Games,"Maplewood, NJ",failed,4000.0,20.0,0.005000,2,"Mon, 02 Aug 2010 03:59:00 -0000",5,"$1,$5,$10,$25,$50",6,0,47.18
2,237090,GETTING OVER - One son's search to finally kno...,http://www.kickstarter.com/projects/charnick/g...,Film & Video,Documentary,"Los Angeles, CA",successful,6000.0,6535.0,1.089167,100,"Sun, 08 Apr 2012 02:14:00 -0000",13,"$1,$10,$25,$30,$50,$75,$85,$100,$110,$250,$500...",4,0,32.22
3,246101,The Launch of FlyeGrlRoyalty &quot;The New Nam...,http://www.kickstarter.com/projects/flyegrlroy...,Fashion,Fashion,"Novi, MI",failed,3500.0,0.0,0.000000,0,"Wed, 01 Jun 2011 15:25:39 -0000",6,"$10,$25,$50,$100,$150,$250",2,0,30.00
4,316217,Dinner Party - a short film about friendship.....,http://www.kickstarter.com/projects/249354515/...,Film & Video,Short Film,"Portland, OR",successful,3500.0,3582.0,1.023331,39,"Wed, 22 Jun 2011 13:33:00 -0000",7,"$5,$25,$50,$100,$250,$500,$1,000",8,0,21.43


In [4]:
# Get the dimension of the dataset
df.shape

(41965, 17)

In [5]:
# Get the summary of the dataset
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 41965 entries, 0 to 41964
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   project id         41965 non-null  int64  
 1   name               41965 non-null  str    
 2   url                41965 non-null  str    
 3   category           41965 non-null  str    
 4   subcategory        41965 non-null  str    
 5   location           40643 non-null  str    
 6   status             41965 non-null  str    
 7   goal               41965 non-null  float64
 8   pledged            41965 non-null  float64
 9   funded percentage  41965 non-null  float64
 10  backers            41965 non-null  int64  
 11  funded date        41965 non-null  str    
 12  levels             41965 non-null  int64  
 13  reward levels      41906 non-null  str    
 14  updates            41965 non-null  int64  
 15  comments           41965 non-null  int64  
 16  duration           41965 non-null

In [6]:
# Check for missing values in the dataset
missing_count = df.isnull().sum()
missing_count[missing_count > 0]

location         1322
reward levels      59
dtype: int64

In [7]:
# Calculate the percentage of missing values for each column to identify which columns have a significant amount of missing data
missing_percentage = (missing_count / len(df)) * 100
missing_percentage[missing_percentage > 0].round(2)

location         3.15
reward levels    0.14
dtype: float64

In [8]:
# Check the first 10 rows with missing values in the "location" column
df[df["location"].isnull()].head(10)

,project id,name,url,category,subcategory,location,status,goal,pledged,funded percentage,backers,funded date,levels,reward levels,updates,comments,duration
9,727286,Offline Wikipedia iPhone app,http://www.kickstarter.com/projects/dphiffer/o...,Technology,Open Software,NaN,successful,99.0,145.0,1.464646,25,"Tue, 14 Jul 2009 06:59:59 -0000",1,$1,5,19,79.64
12,893085,Esperanza Farm: A Novel,http://www.kickstarter.com/projects/JesusMaria...,Publishing,Fiction,NaN,failed,6500.0,765.0,0.117692,20,"Fri, 16 Jul 2010 03:59:00 -0000",7,"$5,$10,$25,$50,$75,$150,$300",10,0,60.34
42,2442649,Dream with Me--A documentary about one year in...,http://www.kickstarter.com/projects/1800556280...,Film & Video,Documentary,NaN,successful,5000.0,5615.0,1.123000,90,"Fri, 16 Oct 2009 01:08:00 -0000",8,"$10,$25,$50,$100,$500,$1,000,$2,500,$5,000",6,7,30.89
85,4719237,Indie Music Website (iTunes for Indie Musicians),http://www.kickstarter.com/projects/binarykoll...,Music,Indie Rock,NaN,failed,10000.0,10.0,0.001000,1,"Thu, 01 Oct 2009 06:19:00 -0000",3,"$1,$20,$100",0,0,29.98
87,4732285,Insiders/Out: Exploring Outsider Art in America,http://www.kickstarter.com/projects/insidersou...,Art,Art,NaN,successful,2000.0,2000.0,0.999830,30,"Thu, 08 Jul 2010 21:39:00 -0000",13,"$1,$5,$10,$15,$20,$25,$35,$50,$75,$100,$150,$2...",4,1,42.97
91,4948516,STRANGE POSITIONING SYSTEMS (SPS) - A global l...,http://www.kickstarter.com/projects/sps/strang...,Art,Performance Art,NaN,failed,8500.0,695.0,0.081765,15,"Fri, 20 Aug 2010 21:37:00 -0000",10,"$10,$25,$50,$75,$100,$250,$500,$600,$1,000,$1,500",3,1,42.19
107,5785733,Faith - a film by Eli Daughdrill,http://www.kickstarter.com/projects/771183968/...,Film & Video,Film &amp; Video,NaN,failed,50000.0,7890.0,0.157800,25,"Fri, 16 Apr 2010 05:48:00 -0000",8,"$10,$25,$50,$100,$250,$500,$1,000,$5,000",8,2,74.39
134,7808555,Fabric Fiction: A collaborative literary exper...,http://www.kickstarter.com/projects/1930061782...,Publishing,Fiction,NaN,failed,3000.0,582.0,0.194000,13,"Fri, 04 Jun 2010 16:30:00 -0000",10,"$1,$5,$10,$12,$20,$25,$40,$50,$75,$100",7,0,89.94
158,9094253,"The People, Places &amp; Patterns Project",http://www.kickstarter.com/projects/pwc/the-pe...,Photography,Photography,NaN,failed,10000.0,3078.0,0.307752,43,"Tue, 16 Feb 2010 19:11:00 -0000",9,"$30,$60,$80,$100,$150,$250,$500,$1,000,$2,000",11,6,61.67
166,9316432,Produce JigGsaw's never before heard &quot;The...,http://www.kickstarter.com/projects/jiggsaw/pr...,Music,Rock,NaN,failed,700.0,25.0,0.035714,1,"Mon, 07 Jun 2010 04:00:00 -0000",6,"$15,$25,$50,$100,$500,$1,000",0,0,89.78


In [9]:
# Verify that the status of the campaigns with missing locations is either successful or failed
df[df["location"].isnull()]["status"].value_counts()

status
failed        692
successful    630
Name: count, dtype: int64

Location has 1,322 missing values (3.15% of the filtered dataset). The affected records include both successful and failed campaigns and still contain useful campaign data. Instead of dropping these rows, missing locations will be labeled as "Unknown" to preserve the records without assuming a location.

In [10]:
# Fill missing locations with "Unknown"
df["location"] = df["location"].fillna("Unknown")

In [11]:
# Verify there are no missing locations left
df["location"].isnull().sum()

np.int64(0)

### Pledged

In the original dataset, there were 12 missing `pledged` values, and all 12 belonged to campaigns with `status = "live"`.

Task #2 filters the dataset to retain only `successful` and `failed` campaigns for the binary classification objective. As a result, the `live` campaigns containing the missing `pledged` values were removed before this missing-value handling step.

The filtered dataset therefore contains no missing `pledged` values, so no imputation is needed.

In [12]:
# Inspect missing reward levels values to see whether related campaign fields reveal a pattern
df[df["reward levels"].isnull()][["name", "levels", "reward levels", "status"]]

,name,levels,reward levels,status
67,DEAR HARVEY (Stories of Harvey Milk)...Educati...,0,NaN,failed
753,Having homeless create public art,0,NaN,failed
2475,The FullCircle Project: Skiing and Service in ...,0,NaN,successful
2790,Gifts and Crafts made from wood(Golf Cart),0,NaN,failed
3531,The Pyrosphere: large scale computer controlle...,0,NaN,failed
4682,Reanimation Library Coffer Builder,0,NaN,successful
7849,ENUR REQUIEM is going back into the studio!!!!!,0,NaN,failed
9691,Back &quot;The Pack&quot;!!!,0,NaN,successful
9767,Ballard High School Debate,0,NaN,successful
9799,THE FUTURE,0,NaN,failed


### Reward Levels

There are 59 missing values in `reward levels` (0.14% of the filtered dataset).

All 59 missing values occur where `levels = 0`, which indicates these campaigns have no reward tiers rather than unknown reward information.

Therefore, the missing values will be replaced with `"No reward levels"`.

In [13]:
# Fill missing reward levels with "No reward levels"
df["reward levels"] = df["reward levels"].fillna("No reward levels")

In [14]:
# Verify there are no missing reward levels left
df["reward levels"].isnull().sum()

np.int64(0)

In [15]:
# Check for any remaining missing values in the dataset
remaining_missing = df.isnull().sum()
remaining_missing[remaining_missing > 0]

Series([], dtype: int64)

In [16]:
# Save the dataset after handling missing values
output_path = "../data/kickstarter_missing_values_handled.csv"

df.to_csv(output_path, index=False)

print(f"Saved cleaned dataset to '{output_path}'")

Saved cleaned dataset to '../data/kickstarter_missing_values_handled.csv'


## Missing Values Summary

After Task #2 filtered the dataset to retain only `successful` and `failed` campaigns, the dataset contains 41,965 rows and 17 columns.

Missing values were identified in two columns:

| Column | Missing Count | Missing % | Strategy |
|---|---:|---:|---|
| location | 1,322 | 3.15% | Replaced with `"Unknown"` |
| reward levels | 59 | 0.14% | Replaced with `"No reward levels"` |

### Decisions

- **Location:** The missing-location records contained valid campaign information and included both successful and failed campaigns. Rather than dropping 3.15% of the filtered dataset, missing locations were labeled `"Unknown"`.

- **Pledged:** In the original dataset, all 12 missing `pledged` values belonged to `live` campaigns. These campaigns were removed during Task #2 when the dataset was filtered to retain only `successful` and `failed` outcomes. Therefore, the filtered dataset contains no missing `pledged` values and no imputation was required.

- **Reward Levels:** All 59 missing `reward levels` values occurred where `levels = 0`, indicating that these campaigns had no reward tiers. These values were replaced with `"No reward levels"`.

After applying these strategies, no missing values remain in the filtered dataset.